### PR TITLE
replace cloneable service with httpflow abstraction

### PR DESCRIPTION
**Pull request by fakeshadow**
## PR Type

Refactor

## PR Checklist
Check your PR fulfills the following:

- [x] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [x] Format code with the latest stable rustfmt

## Overview

Remove `CloneableService` from `actix-http`.
Group `UpgradeHandler`, `ExpectHandler` and HttpService's service type into one `HttpServices` type. This would reduce heap allocation and runtime borrow check.
Add `OnConnectData` type to manage conditional on_connect_data merge in dispatchers. This would reduce the cost of iteration and empty check of it's hashmap.
pin upgrade future in `h1::Dispatcher`.

### CODE REVIEW

File: actix-http/src/service.rs [hunk 1]
```diff
@@ -441,14 +441,29 @@ where
...
-    expect: CloneableService<X>,
-    upgrade: Option<CloneableService<U>>,
+    services: Rc<RefCell<HttpServices<S, X, U>>>,
     cfg: ServiceConfig,
     on_connect_ext: Option<Rc<ConnectCallback<T>>>,
     _phantom: PhantomData<B>,
 }
 
+// a collection of service for http.
+pub(super) struct HttpServices<S, X, U> {
```
Comment (robjtede) [inline actix-http/src/service.rs]: I'd like a more searchable name for it. `HttpFlow`?

File: actix-http/src/lib.rs [hunk 2]
```diff
@@ -81,3 +80,36 @@ pub enum Protocol {
 }
 
 type ConnectCallback<IO> = dyn Fn(&IO, &mut Extensions);
+
+// container for data that extract with ConnectCallback
```
Comment (robjtede) [inline actix-http/src/lib.rs]: use doccomments in this file

```suggestion
/// Container for data that extract with ConnectCallback.
```

File: actix-http/src/lib.rs [hunk 3]
```diff
@@ -81,3 +80,36 @@ pub enum Protocol {
...
+            handler(io, &mut extensions);
+            extensions
+        });
+
+        Self(ext)
+    }
+
+    // merge self to given request's head extension.
+    #[inline]
+    pub(crate) fn merge(&mut self, req: &mut Request) {
```
Comment (robjtede) [inline actix-http/src/lib.rs]: just to make the merge direction clear when reading the dispatcher: `merge_into`, `drain_into`, `drain_into_request`, etc.

```suggestion
    pub(crate) fn merge_into(&mut self, req: &mut Request) {
```

File: actix-http/src/service.rs (hunk 1 above)
Comment (fakeshadow) [reply actix-http/src/service.rs]: np.

**Pull request merged**